dot/parachain/candidate-validation/host.go

```diff
@@ -0,0 +1,103 @@
+package candidatevalidation
+
+import (
+	"fmt"
+
+	parachainruntime "github.com/ChainSafe/gossamer/dot/parachain/runtime"
+	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/internal/log"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+var logger = log.NewFromGlobal(log.AddContext("pkg", "pvf"), log.SetLevel(log.Debug))
+
+type Host struct {
+	workerPool *workerPool
+}
+
+func NewValidationHost() *Host {
+	return &Host{
+		workerPool: newValidationWorkerPool(),
+	}
+}
+
+func (v *Host) Validate(msg *ValidationTask) (*ValidationResult, error) {
+	validationCodeHash := msg.ValidationCode.Hash()
+	// performBasicChecks
+	validationErr, internalErr := performBasicChecks(&msg.CandidateReceipt.Descriptor,
+		msg.PersistedValidationData.MaxPovSize,
+		msg.PoV,
+		validationCodeHash)
+
+	if internalErr != nil {
+		return nil, internalErr
+	}
+	if validationErr != nil {
+		return &ValidationResult{InvalidResult: validationErr}, nil //nolint
+	}
+	// create worker if not in pool
+	if !v.workerPool.containsWorker(validationCodeHash) {
+		worker, err := v.workerPool.newValidationWorker(*msg.ValidationCode)
+		if err != nil {
+			return nil, err
+		}
+
+		// sanity check
+		if worker.workerID != validationCodeHash {
+			return nil, fmt.Errorf("workerID does not match validationCodeHash")
+		}
+	}
+
+	// submit request
+	validationParams := parachainruntime.ValidationParameters{
+		ParentHeadData:         msg.PersistedValidationData.ParentHead,
+		BlockData:              msg.PoV.BlockData,
+		RelayParentNumber:      msg.PersistedValidationData.RelayParentNumber,
+		RelayParentStorageRoot: msg.PersistedValidationData.RelayParentStorageRoot,
+	}
+	workTask := &workerTask{
+		work:             validationParams,
+		maxPoVSize:       msg.PersistedValidationData.MaxPovSize,
+		candidateReceipt: msg.CandidateReceipt,
+	}
+	return v.workerPool.submitRequest(validationCodeHash, workTask)
+}
+
+// performBasicChecks Does basic checks of a candidate. Provide the encoded PoV-block.
+// Returns ReasonForInvalidity and internal error if any.
+func performBasicChecks(candidate *parachaintypes.CandidateDescriptor, maxPoVSize uint32,
+	pov parachaintypes.PoV, validationCodeHash parachaintypes.ValidationCodeHash) (
+	validationError *ReasonForInvalidity, internalError error) {
+	povHash, err := pov.Hash()
+	if err != nil {
+		return nil, fmt.Errorf("hashing PoV: %w", err)
+	}
+
+	encodedPoV, err := scale.Marshal(pov)
+	if err != nil {
+		return nil, fmt.Errorf("encoding PoV: %w", err)
+	}
+	encodedPoVSize := uint32(len(encodedPoV))
+
+	if encodedPoVSize > maxPoVSize {
+		ci := ParamsTooLarge
+		return &ci, nil
```

Comment:
I see that the return should be a pointer to `ReasonForInvalidity`, I would like to ask if there is any reason for that? I would suggest to return the raw value itself, and then the caller uses the reference to it

Reply:
I have this returning a pointer so that the caller can check if the return value is nil, then there was no error, the basic check passes.  However it does feel clumsy to treat this as a pointer and having to de-reference this.  So, it's currently if ReasonForInvalidity is nil, then it is valid.